event-sourcing/event-store-postgres/src/main/java/org/apache/james/eventsourcing/eventstore/postgres/PostgresEventStoreDAO.java

```diff
@@ -0,0 +1,122 @@
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.SNAPSHOT;
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.TABLE_NAME;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.eventsourcing.AggregateId;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.EventId;
+import org.apache.james.eventsourcing.eventstore.History;
+import org.apache.james.util.ReactorUtils;
+import org.jooq.JSON;
+import org.jooq.Record;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.jdk.javaapi.CollectionConverters;
+
+public class PostgresEventStoreDAO {
+    private PostgresExecutor postgresExecutor;
+    private JsonEventSerializer jsonEventSerializer;
+
+    @Inject
+    public PostgresEventStoreDAO(PostgresExecutor postgresExecutor, JsonEventSerializer jsonEventSerializer) {
+        this.postgresExecutor = postgresExecutor;
+        this.jsonEventSerializer = jsonEventSerializer;
+    }
+
+    public Mono<Void> appendAll(List<Event> events, Optional<EventId> lastSnapshot) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, AGGREGATE_ID, EVENT_ID, EVENT)
+                .valuesOfRecords(events.stream().map(event -> dslContext.newRecord(AGGREGATE_ID, EVENT_ID, EVENT)
+                    .value1(event.getAggregateId().asAggregateKey())
+                    .value2(event.eventId().serialize())
+                    .value3(convertToJooqJson(event))).collect(ImmutableList.toImmutableList()))))
+            .then(lastSnapshot.map(eventId -> insertSnapshot(events.iterator().next().getAggregateId(), eventId)).orElse(Mono.empty()));
+    }
+
+    private Mono<Void> insertSnapshot(AggregateId aggregateId, EventId snapshotId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+            .set(SNAPSHOT, snapshotId.serialize())
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))));
+    }
+
+    private JSON convertToJooqJson(Event event) {
+        try {
+            return JSON.json(jsonEventSerializer.serialize(event));
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public Mono<EventId> getSnapshot(AggregateId aggregateId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.select(SNAPSHOT)
+            .from(TABLE_NAME)
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))
+            .limit(1)))
+            .map(record -> EventId.fromSerialized(Optional.ofNullable(record.get(SNAPSHOT)).orElse(0)));
```

Comment:
`Optional.ofNullable(record.get(SNAPSHOT)).orElse(0))`

IMO use optional to represent emptyness, not a special value

Reply:
just a way to create default value in case record.get(SNAPSHOT) is null. Otherwise I have check if record.get(SNAPSHOT) is null

Reply:
No null

No special value

The way to handle absence shall be handled with Optional.

Reply:
In Cassandra implementation we are falling back SNAPSHOT default to 0 if null: https://github.com/apache/james-project/blob/master/event-sourcing/event-store-cassandra/src/main/scala/org/apache/james/eventsourcing/eventstore/cassandra/EventStoreDao.scala#L143. (Cassandra driver while `.getInt` would fallback to 0 if CQL value is null).

I think `Optional.ofNullable(record.get(SNAPSHOT)).orElse(0))` is fine here and aligned with what we did on Cassandra implementation.

(Unless we mistakenly fallbacked SNAPSHOT to 0 in the beginning).